### PR TITLE
Rn validation mp

### PR DESCRIPTION
**Pull request by Noy-Maimon**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: demisto/etc#33160

## Description
Added the option for modified existing release notes, and handled the modification of the bucket index required for these changes. 

## Bucket upload build 
[Green bucket upload build](https://app.circleci.com/pipelines/github/demisto/content/77278/workflows/b6884ea9-8fe8-4c3c-bf76-21f24cac80a3) checking the following cases:
1. A pack that a random version was changed
2. A pack that the latest version was changed 
3. A new pack (different build, [here](https://app.circleci.com/pipelines/github/demisto/content/77764/workflows/e56c98cc-b73d-4f4d-a199-84c16cf86f53))
4. A pack that both latest & random versions were changed, for both aggregated and un-aggregated versions in the changelog file.
5. Change in an RN that is a key in the changelog of an aggregated block.
6. Change in the RN that is in version 1.0.0, and it is not the latest version.
7. Change in the RN that is in version 1.0.0, and it is the latest version.

# Screen Shots
Screen shots of the changelog files under index.zip derived from the test upload.
The actual index.zip for cases 1, 2, 4 can be found [here](https://console.cloud.google.com/storage/browser/marketplace-dist-dev/upload-flow/builds/rn_validation_mp/318784/content/packs?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%257B_22k_22_3A_22_22_2C_22t_22_3A10_2C_22v_22_3A_22_5C_22index_5C_22_22%257D%255D%22))&project=oproxy-dev&prefix=index&forceOnObjectsSortingFiltering=false) (under the packs: `FeedBambenekConsulting`, , `dnstwist` and `Base`)and for cases 5-7 can be found [here](https://console.cloud.google.com/storage/browser/_details/marketplace-ci-build/content/builds/rn_validation_mp/329046/content/packs/index.zip?project=oproxy-dev) (under the packs: `AzureStorage`, `Barracuda` and `Base`).

**case 1:** 
Pack: FeedBambenekConsulting
image

**case 2:**
Pack: dnstwist
image

**case 3:**
Pack: TEST_RN_VALIDATION_DUMMY_PACK
to be added
image

**case 4:**
Pack: Base

part 1/3:
image

part 2/3:
image

part 3/3:
image

**case 5:**
Pack: Base
image

**case 6:**
Pack: Barracuda
image

**case 7:**
Pack: Azure Storage
Screen Shot 2021-04-18 at 16 30 18

## Example for case 4 in XSOAR instance:

image

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [x] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/Marketplace/marketplace_services.py [hunk 1]
```diff
@@ -1233,6 +1235,91 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+        Returns:
+            A dict of version, rn data for all corresponding versions, and the highest version among those keys as str
+
+        """
+        same_block_versions_dict: dict = dict()
+        versions = [LooseVersion(item) for item in changelog.keys()]
+        higher_versions = [item for item in versions if item > version]
+        lower_versions = [item for item in versions if item < version]
+        higher_nearest_version = min(higher_versions)
+        lower_nearest_version = max(lower_versions)
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1312]: This code is iterating on `changelog-keys` 5 times, better to do it in one iteration.

File: Tests/Marketplace/Tests/marketplace_services_test.py [hunk 2]
```diff
@@ -741,6 +755,72 @@ def test_prepare_release_notes_upgrade_version_dup(self, mocker, dummy_pack):
         assert task_status is True
         assert not_updated_build is False
 
+    @pytest.mark.parametrize('version, boolean_value', [('1.0.1', True), ('1.0.2', False)])
```
Comment (guyfreund) [inline Tests/Marketplace/Tests/marketplace_services_test.py:758]: Please test the case where it is in the changelog but it is part of aggregated version

File: Tests/Marketplace/marketplace_services.py (hunk 1 above)
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1312]: If `version=1.0.2` and in changelog we have only `1.0.2` and `1.0.0`, then `1.0.1` was aggregated along with `1.0.2`. So `higher_versions=[]` and you will get `ValueError` using `max`.

Along with fixing this, please check that case in a testing upload

**Review by guyfreund — CHANGES_REQUESTED** [review]
Very nice work! 👏  , a few notes:
1. Please delete the dummy files
2. Please revert the requirements file
3. Have we tested this on XSOAR instance to see how the changes looks on MP?
4. Not sure what every case is testing, so please mention it

**Review by guyfreund — CHANGES_REQUESTED** [review]
- Please also test the edition of version `1.0.0` of a pack. Both when it's the latest version and not the latest version.

File: Tests/Marketplace/marketplace_services.py [hunk 3]
```diff
@@ -2732,3 +2857,31 @@ def is_ignored_pack_file(modified_file_path_parts):
...
+    Args:
+        release_notes_dir: the path to the release notes dir.
+        version (str): the wanted version.
+        changelog (dict): the changelog from the production bucket.
+
+    Returns:
+        True if this version's value in the changelog is not an aggregated release notes block. False otherwise.
+    """
+    if changelog.get(version):
+        all_rn_versions = []
```
[the remark this answers is not in the document]
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py:2874]: Done

File: Tests/Marketplace/Tests/marketplace_services_test.py (hunk 2 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/Tests/marketplace_services_test.py:758]: Done

File: Tests/Marketplace/marketplace_services.py (hunk 1 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py:1309]: Done

File: Tests/Marketplace/marketplace_services.py [hunk 4]
```diff
@@ -2732,3 +2857,31 @@ def is_ignored_pack_file(modified_file_path_parts):
...
+        lowest_version = [LooseVersion('0.0.0')]
+        for filename in sorted(os.listdir(release_notes_dir)):
+            _current_version = filename.replace('.md', '')
+            current_version = _current_version.replace('_', '.')
+            all_rn_versions.append(LooseVersion(current_version))
+        lower_versions_all_versions = [item for item in all_rn_versions if item < version] + lowest_version
+        lower_versions_in_changelog = [LooseVersion(item) for item in changelog.keys() if
+                                       LooseVersion(item) < version] + lowest_version
```
[the remark this answers is not in the document]
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py:2892]: As discussed, left as is for cases such as https://github.com/demisto/etc/issues/35723

File: Tests/Marketplace/marketplace_services.py [hunk 5]
```diff
@@ -1229,6 +1231,86 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+        Returns:
+            A dict of version, rn data for all corresponding versions, and the highest version among those keys as str
+
+        """
+        lower_versions, higher_versions = [], []
+        same_block_versions_dict: dict = dict()
+        for item in changelog.keys():  # divide the versions into lists of lower and higher than given version
+            (lower_versions if LooseVersion(item) < version else higher_versions).append(LooseVersion(item))
+        higher_nearest_version = min(higher_versions)
+        lower_nearest_version = max(lower_versions)
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1303]: Still need to handle edge case of `max` (if pack version is `1.0.0`)

**Review by guyfreund — CHANGES_REQUESTED** [review]
Nice

File: Tests/Marketplace/marketplace_services.py (hunk 5 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py:1305]: Done

**Review by guyfreund — APPROVED** [review]

File: Tests/Marketplace/marketplace_services.py [hunk 6]
```diff
@@ -1229,6 +1231,88 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+
+        """
+
+        modified_versions_dict = {}
+
+        for rn_filename in modified_rn_files:
+            _version = rn_filename.replace('.md', '')
+            version = _version.replace('_', '.')
+            # Should only apply on modified files that are not the last rn file
+            if LooseVersion(version) < changelog_latest_rn_version:
```
Comment (yaakovi) [inline Tests/Marketplace/marketplace_services.py:1265]: reverse the condition to decrease the "area under the indent"

File: Tests/Marketplace/marketplace_services.py [hunk 7]
```diff
@@ -1229,6 +1231,88 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+                # The case where the version is a key in the changelog file,
+                # and the value is not an aggregated release note
+                if is_the_only_rn_in_block(release_notes_dir, version, changelog):
+                    logging.info("The version is a key in the changelog file and by itself in the changelog block")
```
Comment (yaakovi) [inline Tests/Marketplace/marketplace_services.py:1269]:
```suggestion
                    logging.debug(f'The "{version}" version is a key in the changelog file and by itself in the changelog block')
```

File: Tests/Marketplace/marketplace_services.py [hunk 8]
```diff
@@ -1229,6 +1231,88 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+                    with open(os.path.join(release_notes_dir, rn_filename), 'r') as rn_file:
+                        rn_lines = rn_file.read()
+                    modified_versions_dict[version] = self._clean_release_notes(rn_lines).strip()
+                # The case where the version is not a key in the changelog file or it is a key of aggregated content
+                else:
+                    logging.info("The version is not a key in the changelog file or it is a key of aggregated content")
```
Comment (yaakovi) [inline Tests/Marketplace/marketplace_services.py:1275]:
```suggestion
                    logging.debug(f'The "{version}" version is not a key in the changelog file or it is a key of aggregated content')
```

File: Tests/Marketplace/marketplace_services.py [hunk 9]
```diff
@@ -1229,6 +1231,88 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+
+        Returns:
+            A dict of version, rn data for all corresponding versions, and the highest version among those keys as str
+
+        """
+        lowest_version = [LooseVersion(Pack.PACK_INITIAL_VERSION)]
+        lower_versions, higher_versions = [], []
+        same_block_versions_dict: dict = dict()
+        for item in changelog.keys():  # divide the versions into lists of lower and higher than given version
+            (lower_versions if LooseVersion(item) < version else higher_versions).append(LooseVersion(item))
```
Comment (yaakovi) [inline Tests/Marketplace/marketplace_services.py:1302]: nice

File: Tests/Marketplace/marketplace_services.py [hunk 10]
```diff
@@ -1229,6 +1231,88 @@ def get_changelog_latest_rn(self, changelog_index_path: str) -> Tuple[dict, Loos
...
+        higher_nearest_version = min(higher_versions)
+        lower_versions = lower_versions + lowest_version  # if the version is 1.0.0, ensure lower_versions is not empty
+        lower_nearest_version = max(lower_versions)
+        for rn_filename in os.listdir(release_notes_dir):
+            _current_version = rn_filename.replace('.md', '')
+            current_version = _current_version.replace('_', '.')
```
Comment (yaakovi) [inline Tests/Marketplace/marketplace_services.py:1308]: you use this conversion several times:
```
def release_notes_file_to_version(rn_file_name):
    return rn_file_name.replace('.md', '').replace('_', '.')
```
you can also add casting to `LooseVersion` if needed

File: Tests/Marketplace/marketplace_services.py [hunk 11]
```diff
@@ -2728,3 +2850,41 @@ def is_ignored_pack_file(modified_file_path_parts):
...
+
+    Args:
+        release_notes_dir: the path to the release notes dir.
+        version (str): the wanted version.
+        changelog (dict): the changelog from the production bucket.
+
+    Returns:
+        True if this version's value in the changelog is not an aggregated release notes block. False otherwise.
+    """
+    if changelog.get(version):
```
Comment (yaakovi) [inline Tests/Marketplace/marketplace_services.py:2879]: reverse the condition to decrease the "area under the indent"

**Review by yaakovi — APPROVED** [review]
very nice!
see my comments

**Review by guyfreund — COMMENTED** [review]
@Noy-Maimon Please re-trigger a testing upload before merging

File: Tests/Marketplace/marketplace_services.py (hunk 10 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py]: Done

File: Tests/Marketplace/marketplace_services.py [hunk 12]
```diff
@@ -1297,14 +1381,34 @@ def assert_upload_bucket_version_matches_release_notes_version(self,
...
+        """
+        modified_rn_files = []
+        for file_path in modified_files_paths:
+            modified_file_path_parts = os.path.normpath(file_path).split(os.sep)
+            if self.RELEASE_NOTES in modified_file_path_parts:
+                modified_rn_files.append(modified_file_path_parts[-1])
+        return modified_rn_files
+
+    def prepare_release_notes(self, index_folder_path, build_number, pack_was_modified=False,
+                              modified_files_paths: list = []):
```
[the remark this answers is not in the document]
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py]: Done

File: Tests/Marketplace/marketplace_services.py (hunk 6 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py]: Done

File: Tests/Marketplace/marketplace_services.py (hunk 11 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py]: Done
